compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,812 @@
+  -- ^ A daml-lf expression.
+  | BAnd BoolExpr BoolExpr
+  -- ^ And operator.
+  | BNot BoolExpr
+  -- ^ Not operator.
+  deriving Show
+
+-- | Data type denoting a potentially conditional value.
+data Cond a
+  = Determined a
+  -- ^ Non-conditional value.
+  | Conditional BoolExpr a (Maybe a)
+  -- ^ Conditional value, with a (Boolean) condition, a value in case the
+  -- condition holds, and possibly a value in case it doesn't.
+  deriving Show
+
+instance Functor Cond where
+  fmap f (Determined x) = Determined (f x)
+  fmap f (Conditional e x y) = Conditional e (f x) (f <$> y)
+
+-- | Class covering the types which store conditionals inside.
+class Conditional a where
+  -- | Shift the conditional inside the data type.
+  introCond :: Cond a -> a
+
+instance Conditional (UpdateSet ph) where
+  introCond (Determined upds) = upds
+  introCond (Conditional e updx updy) = case updx of
+    UpdateSetVG{} -> UpdateSetVG
+      (concatCond $ Conditional e (_usvgUpdate updx) (_usvgUpdate <$> updy))
+      (concatCond $ Conditional e (_usvgChoice updx) (_usvgChoice <$> updy))
+      (concatCond $ Conditional e (_usvgValue updx) (_usvgValue <$> updy))
+    UpdateSetCG{} -> UpdateSetCG
+      (concatCond $ Conditional e (_uscgUpdate updx) (_uscgUpdate <$> updy))
+      (concatCond $ Conditional e (_uscgChoice updx) (_uscgChoice <$> updy))
+    UpdateSetS{} -> UpdateSetS
+      (concatCond $ Conditional e (_ussUpdate updx) (_ussUpdate <$> updy))
+
+-- | Flatten nested conditionals into a single level.
+concatCond :: Cond [Cond a] -> [Cond a]
+concatCond (Determined xs) = xs
+concatCond (Conditional e xs mys) = case mys of
+  Just ys -> map (ext_cond e) xs ++ map (ext_cond (BNot e)) ys
+  Nothing -> map (ext_cond e) xs
+  where
+    ext_cond :: BoolExpr -> Cond b -> Cond b
+    ext_cond e (Determined x) = Conditional e x Nothing
+    ext_cond e1 (Conditional e2 x y) = Conditional (BAnd e1 e2) x y
+
+-- | Data type denoting an update.
+data Upd
+  = UpdCreate
+  -- ^ Data type denoting a create update.
+    { _creTemp  :: !(Qualified TypeConName)
+     -- ^ Qualified type constructor corresponding to the contract template.
+    , _creField :: ![(FieldName, Expr)]
+      -- ^ The fields to be verified, together with their value.
+    }
+  | UpdArchive
+  -- ^ Data type denoting an archive update.
+    { _arcTemp  :: !(Qualified TypeConName)
+      -- ^ Qualified type constructor corresponding to the contract template.
+    , _arcField :: ![(FieldName, Expr)]
+      -- ^ The fields to be verified, together with their value.
+    }
+  deriving Show
+
+-- | Data type denoting an exercised choice update.
+data UpdChoice = UpdChoice
+  { _choTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _choName  :: !ChoiceName
+    -- ^ The name of the choice.
+  }
+  deriving (Eq, Generic, Hashable, Show)
+
+-- | The collection of updates being performed.
+data UpdateSet (ph :: Phase) where
```

Comment:
My impression is that you're using GADTs here to link the `UpdateSet` types from the three different phases with their `Env` counterparts. This problem might be better solved with associated data types similar to
```haskell
class IsPhase (ph :: Phase) where
  data UpdateSet ph
  data Env ph
```
Let's try that out in a follow-up PR.